### DISCUSSION TITLE
set_view_options seems to have no effect

### DISCUSSION THREAD

**Discussion by lamyseba** [jeremyevans/roda#248] [General]
Hi!
I'm using Roda and trying to use the plugin view_options to change the default behavior when rendering view. Following [the doc](http://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/ViewOptions.html#module-Roda::RodaPlugins::ViewOptions-label-Per-branch+HTML+escaping), I wrote the folowing code:
```
# app.rb

class AdherentWebsite < Roda
  #...
  plugin :render, escape: true, layout: '/layout_admin'
  plugin :view_options
  route do |r|
    #...
    r.hash_routes('')
    #...
  end
end

```
```
# routes/rappel.rb

class AdherentWebsite
  hash_routes.on 'rappels' do |r|
    r.is 'preview' do
      set_view_subdir 'rappels'
      set_view_options(template_opts: {escape: false})
      #...
    end
  end
end
```
set_view_subdir is working ok, but set_view_options has no effect, the output of <%= %> is still "html escaped". Any clue what I might have done wrong ? Is this a bug ?

**Comment by jeremyevans** [comment]
I cannot replicate this issue.  I tried with the following self contained example (in a `config.ru` file):

```ruby
require 'roda'

class AdherentWebsite < Roda
  plugin :render, escape: true, layout: '/layout_admin'
  plugin :view_options
  plugin :hash_routes
  route do |r|
    r.hash_routes('')
  end
end

class AdherentWebsite
  hash_routes.on 'rappels' do |r|
    r.is 'preview' do
      set_view_subdir 'rappels'
      set_view_options(template_opts: {escape: false})
      render(:inline=>"<%= '&' %>\n")
    end
    render(:inline=>"<%= '&' %>\n")
  end
end

run AdherentWebsite
```

Requests for `/rappels` return `&amp;`, and requests for `/rappels/preview` return `&`, showing that automatic escaping is disabled in the preview route, but not in other routes of the rappels branch.

 Can you modify the above self contained example to show the failure case you are experiencing?

**Comment by lamyseba** [comment]
Thank you for your quick answer.  The `render` method does indeed works as expected. Problem is coming when using the `view` method with `set_view_options(template_opts: {escape: false})`, and using `<%= yield %>` in your layout file. With a `<%== yield %>`, it works fine, but this is not what I'm expecting, because setting `plugin :render,  escape: false` does not lead to the same behavior (  in this case, you should  use `<%= yield %>` in your layout file instead of `<%== yield %>`)

**Reply by jeremyevans** [reply]
Seems very strange to use `<%= yield %>` in your layout if the default for your application is `escape: true`.  With `escape: true`, you would almost always want your layout to use `<%== yield %>` However, I think it is possible to support that, as shown in the following example:

```ruby
require 'roda'

class AdherentWebsite < Roda
  plugin :render, escape: true, layout_opts: {:inline=>"<%= yield %>\n", :template_opts=>{:escape=>false}}
  plugin :view_options
  plugin :hash_routes
  route do |r|
    r.hash_routes('')
  end
end

class AdherentWebsite
  hash_routes.on 'rappels' do |r|
    r.is 'preview' do
      set_view_subdir 'rappels'
      set_view_options(template_opts: {escape: false})
      view(:inline=>"<%= '&' %>\n")
    end
    view(:inline=>"<%= '&' %>\n")
  end
end

run AdherentWebsite
```

**Comment by lamyseba** [comment]
This is because I have a special layout file  for this route, that is not the same as the rest of my site. Thought it was a good use case for the `view_options` plugin.

 Your solution passing `{:template_opts=>{:escape=>false}}` has the same effects as setting directly `plugin :render, escape: false` : it makes my special route works with a `<%= yield %>` in the layout file, but breaks all the rest of my site that have a layout file with `<%== yield %>`, because the `escape` option set globally in the render plugin  is not "overridable" in the layout file by the `view_options` plugin settings.

**Reply by jeremyevans** [reply]
Can you use `set_layout_options` in addition to `set_view_options` in the rappels/preview route in order to accomplish what you want?  If not, can you please update the example I posted to show the failure you are experiencing so that I can try to debug?

Note that the example I provided (template_opts inside layout_opts), is not the same as a global `escape: false` option, because the example I provided still enables escaping in all rendered templates (just not the layout).  Nowhere previously in this post did you state that you were using a custom layout in your custom route.  That was a fairly important detail to omit.

**Reply by lamyseba** [reply]
Thank a lot! Combination of
```
set_layout_options(:template_opts=>{:escape=>false})
set_view_options(template_opts: {escape: false})
```
is working to have my view with special layout working all "unescaped", including the layout file. Sorry I didn't mention that I was using a special "unescaped" layout for this route.
